elebumm/RedditVideoMakerBot
# Reddit Post Position Config & Download Progress Bar

### Description
Added:
- Dynamic image clip position for each available background
- Progress bar when downloading youtube video
- Few additional background video (GTA, Rocket League, Assasin Creed) as example

### Why? (Motivation)
Although the bot has produced a lot of good videos with minecraft background, I noticed that on some other background, the image of reddit post was blocking the background (check below image). So, it inspire me to make different configuration for each background. For now, it would be the reddit post image position on the background

<table>
  <tr>
    <td valign="top">
        <img height="200" src="https://i.ibb.co/Yb6NfLZ/Before.png" alt="before">
    </td>
    <td valign="top">
        <img height="200" src="https://i.ibb.co/HqQgDMX/After.png" alt="after">
    </td>
  </tr>
  <tr>
    <td><p align="center">Before</p></td>
    <td><p align="center">After</p></td>
  </tr>
</table>

Also, when trying out different background, I noticed that we need to download all background video simultaneously. Because the time it takes to download all video is quite long, I propose that the bot only need to download for needed video & add the progress bar to track the download progress.

<table>
  <tr>
    <td valign="top">
        <img height="200" src="https://i.ibb.co/CwnFTmK/in-progress.jpg" alt="During">
    </td>
    <td valign="top">
        <img height="200" src="https://i.ibb.co/JHCZpFk/finish.jpg" alt="End">
    </td>
  </tr>
  <tr>
    <td><p align="center">Downloading...</p></td>
    <td><p align="center">Done</p></td>
  </tr>
</table>

### How (?)
For changing the position of reddit post image for each background, I use the available `set_position` method in `moviepy`. After refactoring the code for the background video part, we can further make separate configuration (not only the position of reddit post image), such as the blur effect (opacity) etc. 

For the progress bar, I simply add tqdm progress bar and use `on_progress_callback` that was available from pytube API when downloading the video.

### Test Guide
To make testing the new feature easier, you could follow this step by step guides below (i don't include other functionality test, but it should be the same as other functionality):
1. Testing out different screenshot placement in background.
 - First, might need to read [this](https://zulko.github.io/moviepy/ref/VideoClip/VideoClip.html#moviepy.video.VideoClip.VideoClip.set_position). It contains reference on how moviepy put the image in the background clips
 - Second, try out different background to see the diff. (example: minecraft use `center` , while other like GTA Stunt race move the screenshot slightly higher)
 - Third, feel free to modify the position for each background.
 
<img height="100" src="https://user-images.githubusercontent.com/45505811/177037698-9d82b510-b80c-48bd-b768-56a8e7eb8939.png">

2. Testing out the new progress bar during Youtube video download
- First, run the program with background that you never downloaded (or just remove the background vids)
- Second, wait till the download done. You will see the progress bar during download...

3. Testing out new random background (default, if not filled or unsupported background)
- Don't fill the .env variable `BackgroundChoice`. It will use random video (from supported vids) as background

4. Adding new background (this is not new feature, but this PR slightly alter how to add new video as background)

<img height="200" src="https://user-images.githubusercontent.com/45505811/177037910-ff9e176e-4629-4d16-86ea-dec2e379d5dd.png">
